### DISCUSSION TITLE
Passing arguments to a prisma function call

### DISCUSSION THREAD

**Discussion by manan30** [prisma/prisma#19167] [Q&A]
Hey 👋🏽 

So I have this scenario where I am fetching a project by its `id` and have an optional `args` parameter that allows either `select` or `include`

```ts
async findProjectById(
    id: string,
    args?: Omit<Prisma.ProjectFindUniqueOrThrowArgs, 'where'>,
  ) {
    return this.prismaService.project.findUniqueOrThrow({
      where: { id },
      ...args,
    });
  }
```
In the above snippet if I `include` another model from a different function as shown below prisma does not return the included model
```ts
await findProjectById("1233", { include: { user: true })
```

but if I change the prisma call to this
```ts
async findProjectById(
    id: string,
    args?: Omit<Prisma.ProjectFindUniqueOrThrowArgs, 'where'>,
  ) {
    return this.prismaService.project.findUniqueOrThrow({
      where: { id },
     include: { user: true }
    });
  }
```
I am getting the user model back

Anyone know what might be going on?

**Comment by ajmnz** [comment]
I couldn't satisfy the `findUniqueOrThrow` args in an elegant way, but here's a quick one

```ts
async findProjectById<T extends Omit<Prisma.ProjectFindUniqueOrThrowArgs, 'where'>>(
  id: string,
  args?: T,
): Promise<Prisma.ProjectGetPayload<T>> {
  return (await this.prismaService.project.findUniqueOrThrow({
    where: { id },
    ...args,
  })) as Prisma.ProjectGetPayload<T>;
}
```

The key is to use [Generics](https://www.typescriptlang.org/docs/handbook/2/generics.html)

**Comment by manan30** [comment]
@ajmnz thanks for the answer!

So, actually, the thing that is happening over here is if I run the following snippet below

```ts
async findProjectById(
    id: string,
    args?: Omit<Prisma.ProjectFindUniqueOrThrowArgs, 'where'>,
  ) {
    return this.prismaService.project.findUniqueOrThrow({
      where: { id },
      ...args,
    });
  }

findProjectById("id", { include: { user: true } })
```
it outputs
```ts
{
    id: 'project-id',
    name: 'hello',
    userId: 'user-id'
}
```
which is missing the `user` object in the output that I had included by doing `{ user: true }`
But if I do this 
```ts
async findProjectById(
    id: string
  ) {
    return this.prismaService.project.findUniqueOrThrow({
      where: { id },
     { include: { user: true } }
    });
  }
 
findProjectById("id")
```
it outputs
```ts
{
    id: 'project-id',
    name: 'hello',
    userId: 'user-id',
    user: {
     name: "something"
  }
}
```
So the user object does get included when I pass it directly to the `find` function but not when I pass it externally from another function

Just wanted to clarify this is not a type issue as the `args` does get the correct types without the generics, this is more of an issue with the data that is returned

**Reply by ajmnz** [reply]
Oh I see, sorry for the confusion.

This looks weird, since both ways should yield the same result. Could you share what `prismaService` looks like?

---

By the way, in the example above, that is not a valid object

```ts
{
   where: { id },
   { include: { user: true } }
}

// Should look like this

{
   where: { id },
   include: { user: true }
}
```

**Reply by ludralph** [reply]
Hey @manan30,

it would be really helpful if you could provide the additional information that @ajmnz asked about last time. If your problem is resolved by now or somehow doesn't feel relevant anymore, you can simply close the Discussion as well.
